opensciencegrid/contact-sync/topology_utils.py

```diff
@@ -0,0 +1,375 @@
+    """
+    Generate a dictionary mapping from the VO name (key) to the
+    VO ID (value).
+    """
+    old_no_proxy = os.environ.pop('no_proxy', None)
+    os.environ['no_proxy'] = '.opensciencegrid.org'
+
+    url = update_url_hostname("https://my.opensciencegrid.org/vosummary"
+                              "/xml?all_vos=on&active_value=1", args)
+    if session is None:
+        with get_auth_session(args) as session:
+            response = session.get(url)
+    else:
+        response = session.get(url)
+
+    if old_no_proxy is not None:
+        os.environ['no_proxy'] = old_no_proxy
+    else:
+        del os.environ['no_proxy']
+
+    if response.status_code != requests.codes.ok:
+        raise Exception("MyOSG request failed (status %d): %s" % \
+              (response.status_code, response.text[:2048]))
+
+    root = ET.fromstring(response.content)
+    if root.tag != 'VOSummary':
+        raise Exception("MyOSG returned invalid XML with root tag %s" % root.tag)
+    vo_map = {}
+    for child_vo in root:
+        if child_vo.tag != "VO":
+            raise Exception("MyOSG returned a non-VO  (%s) inside VO summary." % \
+                            root.tag)
+        vo_info = {}
+        for child_info in child_vo:
+            vo_info[child_info.tag] = child_info.text
+        if 'ID' in vo_info and 'Name' in vo_info:
+            vo_map[vo_info['Name'].lower()] = vo_info['ID']
+
+    return vo_map
+
+
+SERVICE_IDS = {'ce': 1,
+               'srmv2': 3,
+               'gridftp': 5,
+               'xrootd': 142,
+               'perfsonar-bandwidth': 130,
+               'perfsonar-latency': 130,
+               'gums': 101,
+              }
+def mangle_url(url, args, session=None):
+    """
+    Given a MyOSG URL, switch to using the hostname specified in the
+    arguments
+    """
+    if not args.host:
+        return url
+    url_list = list(urlparse.urlsplit(url))
+    url_list[1] = args.host
+
+    qs_dict = urlparse.parse_qs(url_list[3])
+    qs_list = urlparse.parse_qsl(url_list[3])
+
+    if getattr(args, 'provides_service', None):
+        if 'service' not in qs_dict:
+            qs_list.append(("service", "on"))
+        for service in args.provides_service.split(","):
+            service = service.strip().lower()
+            service_id = SERVICE_IDS.get(service)
+            if not service_id:
+                raise Exception("Requested service %s not known; known service"
+                                " names: %s" % (service, ", ".join(SERVICE_IDS)))
+            qs_list.append(("service_sel[]", str(service_id)))
+
+    if getattr(args, 'owner_vo', None):
+        vo_map = get_vo_map(args, session)
+        if 'voown' not in qs_dict:
+            qs_list.append(("voown", "on"))
+        for vo in args.owner_vo.split(","):
+            vo = vo.strip().lower()
+            vo_id = vo_map.get(vo)
+            if not vo_id:
+                raise Exception("Requested owner VO %s not known; known VOs: %s" \
```

Comment:
We might want to consider using a more specific exception type here, so that it can be handled appropriately